File: lib/py/union_find.py
# Disjoint Set Union (Union-Find)
class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.size = [1] * (n + 1)
        
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x]) # 경로 압축
        return self.parent[x]
    
    def unite(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)
        
        if root_a == root_b:
            return False
        
        # Union by Size
        if self.size[root_a] < self.size[root_b]:
            root_a, root_b = root_b, root_a
            
        self.parent[root_b] = root_a
        self.size[root_a] += self.size[root_b]
        return True
    
    def same(self, a, b):
        return self.find(a) == self.find(b)
        
    def get_size(self, x):
        return self.size[self.find(x)]

File: lib/cpp/union_find.cpp
#include <numeric>
#include <vector>


using namespace std;

// Disjoint Set Union (Union-Find)
struct DSU {
  vector<int> parent;
  vector<int> size; // 그룹 크기 추적

  DSU(int n) {
    parent.resize(n + 1);
    iota(parent.begin(), parent.end(), 0); // 0, 1, 2, ...
    size.assign(n + 1, 1);
  }

  int find(int x) {
    if (parent[x] == x)
      return x;
    return parent[x] = find(parent[x]); // 경로 압축
  }

  bool unite(int a, int b) {
    a = find(a);
    b = find(b);

    if (a == b)
      return false;

    // Union by Size (더 큰 쪽으로 합치기)
    if (size[a] < size[b])
      swap(a, b);
    parent[b] = a;
    size[a] += size[b];
    return true;
  }

  bool same(int a, int b) { return find(a) == find(b); }

  int getSize(int x) { return size[find(x)]; }
};
